File: app/src/main/java/com/gameautopilot/app/core/ActionRing.kt
package com.gameautopilot.app.core

class ActionRing(private val capacity: Int = 12) {
    private val buf = ArrayDeque<String>(capacity)

    @Synchronized
    fun add(label: String) {
        if (buf.size >= capacity) buf.removeFirst()
        buf.addLast(label)
    }

    @Synchronized
    fun snapshot(): List<String> = buf.toList()

    @Synchronized
    fun clear() {
        buf.clear()
    }
}

File: mac/Sources/GameAutopilotMac/Core/ActionRing.swift
import Foundation

/// Fixed-capacity ring buffer of recent action labels shown to the brain
/// as "recent actions" context. Mirrors core/ActionRing.kt.
final class ActionRing {
    private let capacity: Int
    private var buffer: [String] = []
    private let lock = NSLock()

    init(capacity: Int = 12) {
        self.capacity = capacity
    }

    func add(_ label: String) {
        lock.lock(); defer { lock.unlock() }
        if buffer.count >= capacity { buffer.removeFirst() }
        buffer.append(label)
    }

    func snapshot() -> [String] {
        lock.lock(); defer { lock.unlock() }
        return buffer
    }

    func clear() {
        lock.lock(); defer { lock.unlock() }
        buffer.removeAll()
    }
}
